include/gunrock/applications/pr.hxx

```diff
@@ -0,0 +1,200 @@
+namespace pr {
+
+template <typename weight_t>
+struct param_t {
+  weight_t alpha;
+  weight_t tol;
+  
+  param_t(weight_t _alpha, weight_t _tol) : alpha(_alpha), tol(_tol) {}
+};
+
+template <typename weight_t>
+struct result_t {
+  weight_t* p;
+  result_t(weight_t* _p)
+      : p(_p) {}
+};
+
+template <typename graph_t, typename param_type, typename result_type>
+struct problem_t : gunrock::problem_t<graph_t> {
+  param_type param;
+  result_type result;
+
+  problem_t(graph_t& G,
+            param_type& _param,
+            result_type& _result,
+            std::shared_ptr<cuda::multi_context_t> _context)
+      : gunrock::problem_t<graph_t>(G, _context),
+        param(_param),
+        result(_result) {}
+
+  using vertex_t = typename graph_t::vertex_type;
+  using edge_t = typename graph_t::edge_type;
+  using weight_t = typename graph_t::weight_type;
+
+  thrust::device_vector<weight_t> plast;
+  thrust::device_vector<weight_t> idegrees;
+
+  void init() {
+    auto g = this->get_graph();
+    auto n_vertices = g.get_number_of_vertices();
+    plast.resize(n_vertices);
+    idegrees.resize(n_vertices);
+  }
+
+  void reset() {
+    auto g = this->get_graph();
+    auto n_vertices = g.get_number_of_vertices();
+
+    thrust::fill_n(thrust::device, this->result.p, n_vertices, 1.0 / n_vertices);
+    thrust::fill_n(thrust::device, plast.begin(), n_vertices, 0);
+        
+    weight_t alpha = 0.85;
+    auto get_degree = [=] __device__ (const int& i) -> weight_t {
+      return alpha / g.get_number_of_neighbors(i);
+    };
+
+    thrust::transform(
+      thrust::device,
+      thrust::counting_iterator<vertex_t>(0), 
+      thrust::counting_iterator<vertex_t>(n_vertices),
+      idegrees.begin(),
+      get_degree
+    );
+  }
+};
+
+template <typename problem_t>
+struct enactor_t : gunrock::enactor_t<problem_t> {
+  // Use Base class constructor -- does this work? does it handle copy
+  // constructor?
+  using gunrock::enactor_t<problem_t>::enactor_t;
+
+  using vertex_t = typename problem_t::vertex_t;
+  using edge_t = typename problem_t::edge_t;
+  using weight_t = typename problem_t::weight_t;
+
+  void prepare_frontier(frontier_t<vertex_t>* f, cuda::multi_context_t& context) override {
+    auto P = this->get_problem();
+    auto G = P->get_graph();
+    auto n_vertices = G.get_number_of_vertices();
+    
+    for (vertex_t v = 0; v < n_vertices; ++v)
+      f->push_back(v);
+  }
+
+  void loop(cuda::multi_context_t& context) override {
+    // Data slice
+    auto E = this->get_enactor();
+    auto P = this->get_problem();
+    auto G = P->get_graph();
+
+    auto n_vertices = G.get_number_of_vertices();
+    auto p          = P->result.p;
+    auto plast      = P->plast.data().get();
+    auto idegrees   = P->idegrees.data().get();
+    auto alpha      = P->param.alpha;
+    
+    thrust::copy_n(thrust::device, p, n_vertices, plast);
+    thrust::fill_n(thrust::device, p, n_vertices, (1 - alpha) / n_vertices);
```

Comment:
Same comments about the context and the stream. I may just make `context->execution_policy()` which will return the execution policy that goes in place with `thrust::device`.